### ISSUE TITLE
Magento2.4.7 Graphql products aggregation error

### ISSUE THREAD

**Issue by lano-vargas** [magento/magento2#38998]
### Preconditions and environment

- Magento version: 2.4.7
- Anything else that would help a developer reproduce the bug

### Steps to reproduce

```
{
  products(filter: {color: {eq: "52"}}, sort: {name: DESC}) {
    aggregations {
      attribute_code
      count
      label
      options {
        label
        value
        count
      }
    }
    items {
      name
      sku
      price_range {
        minimum_price {
          regular_price {
            value
            currency
          }
        }
      }
    }
    page_info {
      page_size
    }
  }
}
```

### Expected result

Not to return error:
```
"path": [
                "products",
                "aggregations"
            ]
```
aggregations to return data but it's null 

### Actual result

```
{
    "errors": [
        {
            "message": "Internal server error",
            "locations": [
                {
                    "line": 3,
                    "column": 5
                }
            ],
            "path": [
                "products",
                "aggregations"
            ]
        }
    ],
    "data": {
        "products": {
            "aggregations": null,
            "items": [
                {
                    "name": "Tie Gift",
                    "sku": "I03966",
                    "price_range": {......
```

### Additional information

Upon checking exception log: 

```
GraphQL (3:5)
2:   products(filter: {brand: {eq: "409"}}, sort: {name: DESC}) {
3:     aggregations {
       ^
4:       attribute_code
 {"exception":"[object] (GraphQL\\Error\\Error(code: 0): Warning: Undefined array key \"attribute_type\" in /var/www/vhosts/test.com/htdocs/vendor/magento/module-catalog-graph-ql/DataProvider/Product/LayeredNavigation/Builder/Attribute.php on line 100 at /var/www/vhosts/test.com/htdocs/vendor/webonyx/graphql-php/src/Error/Error.php:170)
[previous exception] [object] (Exception(code: 0): Warning: Undefined array key \"attribute_type\" in /var/www/vhosts/test.com/htdocs/vendor/magento/module-catalog-graph-ql/DataProvider/Product/LayeredNavigation/Builder/Attribute.php on line 100 at /var/www/vhosts/test.com/htdocs/vendor/magento/framework/App/ErrorHandler.php:62)"} []test.com
```

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by lano-vargas** [comment]
@magento give me 2.4-develop instance

**Comment by lano-vargas** [comment]
What is this version Magento ver. dev-develop. The issue is replicable on 2.4.7 can you give version 2.4.7 ? The issue might be fixed on your `dev-develop` version

**Comment by lano-vargas** [comment]
https://github.com/magento give me 2.4.7 instance

**Comment by engcom-Bravo** [comment]
@magento give me 2.4-develop instance

**Comment by engcom-Bravo** [comment]
Hi @lano-vargas,

Thanks for your reporting and collaboration.

We have verified the issue in Latest 2.4-develop instance and the issue is not reproducible.Kindly refer the screenshots.


<img width="1440" alt="Screenshot 2024-08-06 at 09 56 07" src="https://github.com/user-attachments/assets/3cb418ab-852f-41c8-b19c-354073d47455">


We are not getting any aggregation error.It might be possible issue got fixed in Latest 2.4-develop instance.

Thanks.

**Comment by lano-vargas** [comment]
Something weird going on here. It's a specific scenario issue here, I'm not entirely sure how to completely replicate this but I have done is mess about the attribute+options creation/deletion and products attribute options assignment creation/deletion 

It appears that when an attribute option is deleted but there are still products assigned to it the graphl returns some extra array of object which appear to be the product that were assigned to previous attribute that was deleted. 

It return the attribute id on both field e.g:
```
{
       "count": 1,
       "label": "417",
        "value": "417",
        "__typename": "AggregationOption"
}
```

in Label it should return a string name such as `colour` etc...

**Comment by engcom-Bravo** [comment]
@magento give me 2.4-develop instance

**Comment by engcom-Bravo** [comment]
Hi @lano-vargas,

Thanks for your update.

We have verified the issue in Latest 2.4-develop instance and the issue is not reproducible.Kindly refer the screenshots.

We have created attribute and assigned to some of the products and after that deleted that attribute.

<img width="1438" alt="Screenshot 2024-08-27 at 11 53 54" src="https://github.com/user-attachments/assets/98217630-062d-4f26-b14f-42a671314247">



There was no error.We are getting aggregations.

Hence we are closing this issue.

Thanks.

**Comment by hnsr** [comment]
Just wanted to chime in and let you now that I am also observing this exact same error on a 2.4.7 instance. Will try and see if I can help with the reproduction.

**Comment by hnsr** [comment]
Looks like this error was happening due to certain products having non-existent option ID references for the `color` attribute. This was probably the result of a botched data migration in our clients project.

**Comment by raoulguillermo** [comment]
I have(or had) the same problem. What fixed it for me was changing the core code. 
/vendor/magento/module-catalog-graph-ql/DataProvider/Product/LayeredNavigation/Builder/Attribute.php
line 100

replace function build with

  public function build(AggregationInterface $aggregation, ?int $storeId): array
    {
        $attributeOptions = $this->getAttributeOptions($aggregation, $storeId);

        // build layer per attribute
        $result = [];
        foreach ($this->getAttributeBuckets($aggregation) as $bucket) {
            $bucketName = $bucket->getName();
            $attributeCode = \preg_replace('~_bucket$~', '', $bucketName);
            $attribute = $attributeOptions[$attributeCode] ?? [];

            $result[$bucketName] = $this->layerFormatter->buildLayer(
                $attribute['attribute_label'] ?? $bucketName,
                \count($bucket->getValues()),
                $attribute['attribute_code'] ?? $bucketName,
                isset($attribute['position']) ? $attribute['position'] : null
            );

            $options = $this->getSortedOptions($bucket, isset($attribute['options']) ? $attribute['options'] : []);
            foreach ($options as $option) {
                $result[$bucketName]['options'][] = $this->layerFormatter->buildItem(
                    $option['label'],
                    $option['value'],
                    $option['count']
                );
            }
        }

        return $result;
    }

**Comment by raoulguillermo** [comment]
I have(had) the same issue.
Fixed it by changing the code. 

/vendor/magento/module-catalog-graph-ql/DataProvider/Product/LayeredNavigation/Builder/Attribute.php
line 100
 
```php
public function build(AggregationInterface $aggregation, ?int $storeId): array
{
    $attributeOptions = $this->getAttributeOptions($aggregation, $storeId);
    $result = [];
    foreach ($this->getAttributeBuckets($aggregation) as $bucket) {
        $bucketName = $bucket->getName();
        $attributeCode = \preg_replace('~_bucket$~', '', $bucketName);
        $attribute = $attributeOptions[$attributeCode] ?? [];

        $result[$bucketName] = $this->layerFormatter->buildLayer(
            $attribute['attribute_label'] ?? $bucketName,
            \count($bucket->getValues()),
            $attribute['attribute_code'] ?? $bucketName,
            isset($attribute['position']) ? $attribute['position'] : null
        );

        $options = $this->getSortedOptions($bucket, isset($attribute['options']) ? $attribute['options'] : []);
        foreach ($options as $option) {
            $result[$bucketName]['options'][] = $this->layerFormatter->buildItem(
                $option['label'],
                $option['value'],
                $option['count']
            );
        }
    }

    return $result;
}
``` 

I don't think this is the way to go. But works for now

**Comment by LucianTuriacArnia** [comment]
I had the same issue due to a "date" type attribute.

When you get:

vendor/magento/module-catalog-graph-ql/DataProvider/Product/LayeredNavigation/Builder/Attribute.php:85
`$attributeOptions = $this->getAttributeOptions($aggregation, $storeId);`

In this function, you will find that only  \'boolean\' and \'price\' attributes will be returned. My solution was to add "date" type.

vendor/magento/module-catalog-graph-ql/DataProvider/Product/LayeredNavigation/AttributeOptionProvider.php:115
`if (!empty($attributeCodes)) {
            $select->orWhere(
                'a.attribute_code in (?) AND a.frontend_input in (\'boolean\', \'price\', **\'date\'**)',
                $attributeCodes
            );
        }`

**Comment by ancyeaji** [comment]
Do reindexing. This fixed my issue.

**Comment by coresh** [comment]
Helped:

1. Code: no changes
2. Restart varnish cache server
3. Increasing: [redis parameters](https://magento.stackexchange.com/questions/296730/redis-session-error-successfully-broke-lock-for-id-sess-ebfc12b22fa8c8e1fb81dca)

Thank you

**Comment by zexperto** [comment]
@lano-vargas  I have the same issue after upgrading from 2.4.6 to 2.4.7-p4 and the issue happened when I have Select Attribute with 0 value in this table "catalog_product_entity_int" and the attribute type is "select"
after deleting all values with value = 0 for all select attributes, the issue is gone

**Comment by dudzio12** [comment]
So, there are some secured array values in the core like:
`$attribute['attribute_label'] ?? $bucketName`
`$attribute['attribute_code'] ?? $bucketName`

But there are two places that are not:
`$optionLabels = $attribute['attribute_type'] === 'boolean'`
` if (self::ATTRIBUTE_OPTIONS_ONLY_WITH_RESULTS === $attribute['is_filterable']) {`

Patch below:

```patch
diff --git a/vendor/magento/module-catalog-graph-ql/DataProvider/Product/LayeredNavigation/Builder/Attribute.php b/vendor/magento/module-catalog-graph-ql/DataProvider/Product/LayeredNavigation/Builder/Attribute.php
index aca276f27..6e65fb1a0 100644
--- a/vendor/magento/module-catalog-graph-ql/DataProvider/Product/LayeredNavigation/Builder/Attribute.php
+++ b/vendor/magento/module-catalog-graph-ql/DataProvider/Product/LayeredNavigation/Builder/Attribute.php
@@ -97,11 +97,11 @@ class Attribute implements LayerBuilderInterface
                 $attribute['attribute_code'] ?? $bucketName,
                 isset($attribute['position']) ? $attribute['position'] : null
             );
-            $optionLabels = $attribute['attribute_type'] === 'boolean'
+            $optionLabels = (isset($attribute['attribute_type']) && $attribute['attribute_type'] === 'boolean')
                 ? $this->YesNo->toArray()
                 : $attribute['options'] ?? [];
             $result[$bucketName]['options'] = $this->getSortedOptions($bucket, $optionLabels);
-            if (self::ATTRIBUTE_OPTIONS_ONLY_WITH_RESULTS === $attribute['is_filterable']) {
+            if (isset($attribute['is_filterable']) && self::ATTRIBUTE_OPTIONS_ONLY_WITH_RESULTS === $attribute['is_filterable']) {
                 $result[$bucketName]['options'] = array_filter(
                     $result[$bucketName]['options'],
                     fn ($option) => $option['count']
```

**Comment by Polar77** [comment]
@engcom-Bravo 
Please reopen this issue and take into account the last comment with the patch. You have obvious errors in the code.

**Comment by Seppo-Konttinen-Solteq** [comment]
Hi.

I was also able to reproduce this with 2.4.7-p5. Even this is kind of data issue, I think additional checks for the array keys would be good.

So we had custom attribute (select) added to product which is also in attribute set but that attribute didn't have any options. I added one option for testing purposes but that wasn't doing the trick. Then I added to one product that option and after that error was gone.

In `vendor/magento/module-catalog-graph-ql/DataProvider/Product/LayeredNavigation/Builder/Attribute.php`

on line 92, the whole `$attribute` array is empty, thus check on line 100 fails.
